File: android/src/main/java/dev/ashu/capacitor/statusbar/safearea/SafeAreaPlugin.java
package dev.ashu.capacitor.statusbar.safearea;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "SafeArea")
public class SafeAreaPlugin extends Plugin {

    private final SafeArea implementation = new SafeArea();

    @PluginMethod
    public void getStatusBarHeight(PluginCall call) {
        JSObject ret = implementation.getStatusBarHeight(getActivity());
        if (ret != null) {
            call.resolve(ret);
        } else {
            call.reject("Status bar height not obtained");
        }
    }

    @PluginMethod
    public void getSafeAreaInsets(PluginCall call) {
        call.resolve(implementation.getSafeAreaInsets(getActivity()));
    }
}

File: ios/Sources/SafeAreaPlugin/SafeAreaPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(SafeAreaPlugin)
public class SafeAreaPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SafeAreaPlugin"
    public let jsName = "SafeArea"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "echo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getStatusBarHeight", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSafeAreaInsets", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = SafeArea()

    @objc func getStatusBarHeight(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve([
                "height": self.implementation.getStatusBarHeight()
            ])
        }
    }

    @objc func getSafeAreaInsets(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve(self.implementation.getSafeAreaInsets())
        }
    }
}
